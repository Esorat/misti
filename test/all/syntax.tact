//
// Checks if the tools process various Tact syntax constructions.
//

trait T {
  abstract get fun getter(): Int;
  abstract const something: Int;
}
contract TestContract with T {
  override get fun getter(): Int { return 0 }
  override const something: Int = 42;
}
trait Z {
  a: Int;
  get fun getter(): Int {
      return self.a + 1;
  }
}
contract TestContractF {
    a: Int = 42;
    fun test() {}
}
trait TParent {
  a: Int;
  fun getA(): Int {
      return self.a;
  }
}
trait TChild with TParent {
  a: Int; // Declare 'a' to satisfy TParent's requirement
}
contract TestContractT with TChild {
    a: Int = 19;
    fun test(): Int { return self.getA(); }
    external() {}
}

fun test_try(a: Int) {
    try { a += 1; }
    try { /* empty */ }
    try { a = 19; } catch (err) { dump(err); }
}

fun test_loops() {
    let sum: Int = 0;
    let i: Int = 0;
    while (i < 10) {
      i = i + 1;
      sum = sum + i;
    }
    do {
      i = i - 1;
      sum = sum + i;
    } until (i > 10);
    repeat (10) {
      i = i + 1;
      sum = sum + i;
    }
    let m: map<Int, Int> = emptyMap();
    foreach (k, v in m) {}
}

//
// Tact code covering all EffectFlags for testing CallGraph effects.
//
// Effects covered:
// - CALLS_SEND
// - CONTRACT_STATE_READ
// - CONTRACT_STATE_WRITE
// - ACCESSES_DATETIME
// - RANDOMNESS_USE
// - RANDOMNESS_SEED_INITIALIZATION
//

contract EffectTestContract {
    // State variable for read/write, initialized to 0
    someVariable: Int = 0;
    destAddress: Address;
    randomnessSeed: Int = 0;

    // Initialize destAddress in the constructor
    init() {
        self.destAddress = sender();
    }

    // Function that calls 'send' (CALLS_SEND)
    fun funcWithSend() {
        let amount: Int = 100;
        send(SendParameters{
            to: self.destAddress,
            value: amount
        });
    }

    // Function that reads from contract state (CONTRACT_STATE_READ)
    fun funcWithStateRead() {
        let value: Int = self.someVariable;
    }

    // Function that writes to contract state (CONTRACT_STATE_WRITE)
    fun funcWithStateWrite() {
        self.someVariable = 42;
    }

    // Function that accesses datetime (ACCESSES_DATETIME)
    fun funcWithDatetimeAccess() {
        let currentTime: Int = now();
    }

    // Function that uses randomness (RANDOMNESS_USE)
    fun funcWithRandomnessUse() {
        let randValue: Int = random(1, 100);
    }

    // Function that initializes randomness seed (RANDOMNESS_SEED_INITIALIZATION)
    fun funcWithRandomnessSeedInit(seed: Int) {
        self.randomnessSeed = seed;
    }

    fun generatePseudoRandom(): Int {
        self.randomnessSeed = (self.randomnessSeed * 1103515245 + 12345) % pow(2, 31);
        return self.randomnessSeed;
    }

    // Function that combines multiple effects
    fun funcWithMultipleEffects() {
        self.someVariable += random(1, 100);
        send(SendParameters{
            to: self.destAddress,
            value: self.someVariable
        });
        let time: Int = now();
        self.funcWithRandomnessSeedInit(time);
    }
}

fun testTryCatch(a: Int) {
    try {
        a += 1;
    } catch (err) {
        dump(err);
    }
}

fun testLoops() {
    let sum: Int = 0;
    let i: Int = 0;

    while (i < 10) {
        i = i + 1;
        sum = sum + i;
    }

    do {
        i = i - 1;
        sum = sum + i;
    } until (i <= 0);

    repeat (10) {
        i = i + 1;
        sum = sum + i;
    }

    let m: map<Int, Int> = emptyMap();
    m.set(1, 10);
    m.set(2, 20);
    m.set(3, 30);

    foreach (key, value in m) {
        sum = sum + value;
    }
}
