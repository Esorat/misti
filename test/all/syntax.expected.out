[MEDIUM] NeverAccessedVariables: Variable value is never accessed
test/all/syntax.tact:97:9:
  96 |     fun funcWithStateRead() {
> 97 |         let value: Int = self.someVariable;
               ^
  98 |     }
Help: Consider removing the variable
See: https://nowarp.io/tools/misti/docs/detectors/NeverAccessedVariables

[MEDIUM] NeverAccessedVariables: Variable currentTime is never accessed
test/all/syntax.tact:107:9:
  106 |     fun funcWithDatetimeAccess() {
> 107 |         let currentTime: Int = now();
                ^
  108 |     }
Help: Consider removing the variable
See: https://nowarp.io/tools/misti/docs/detectors/NeverAccessedVariables

[MEDIUM] NeverAccessedVariables: Variable randValue is never accessed
test/all/syntax.tact:112:9:
  111 |     fun funcWithRandomnessUse() {
> 112 |         let randValue: Int = random(1, 100);
                ^
  113 |     }
Help: Consider removing the variable
See: https://nowarp.io/tools/misti/docs/detectors/NeverAccessedVariables

[INFO] DumpIsUsed: Found `dump` usage
test/all/syntax.tact:41:35:
  40 |     try { /* empty */ }
> 41 |     try { a = 19; } catch (err) { dump(err); }
                                         ^
  42 | }
Help: Using `dump` in production code can sometimes indicate complex code that requires additional review
See: https://nowarp.io/tools/misti/docs/detectors/DumpIsUsed

[INFO] DumpIsUsed: Found `dump` usage
test/all/syntax.tact:141:9:
  140 |     } catch (err) {
> 141 |         dump(err);
                ^
  142 |     }
Help: Using `dump` in production code can sometimes indicate complex code that requires additional review
See: https://nowarp.io/tools/misti/docs/detectors/DumpIsUsed

[INFO] PreferAugmentedAssign: Prefer augmented assignment: i += 1
test/all/syntax.tact:48:7:
  47 |     while (i < 10) {
> 48 |       i = i + 1;
             ^
  49 |       sum = sum + i;
Help: Consider using augmented assignment instead: i += 1
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: sum += i
test/all/syntax.tact:49:7:
  48 |       i = i + 1;
> 49 |       sum = sum + i;
             ^
  50 |     }
Help: Consider using augmented assignment instead: sum += i
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: i -= 1
test/all/syntax.tact:52:7:
  51 |     do {
> 52 |       i = i - 1;
             ^
  53 |       sum = sum + i;
Help: Consider using augmented assignment instead: i -= 1
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: sum += i
test/all/syntax.tact:53:7:
  52 |       i = i - 1;
> 53 |       sum = sum + i;
             ^
  54 |     } until (i > 10);
Help: Consider using augmented assignment instead: sum += i
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: i += 1
test/all/syntax.tact:56:7:
  55 |     repeat (10) {
> 56 |       i = i + 1;
             ^
  57 |       sum = sum + i;
Help: Consider using augmented assignment instead: i += 1
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: sum += i
test/all/syntax.tact:57:7:
  56 |       i = i + 1;
> 57 |       sum = sum + i;
             ^
  58 |     }
Help: Consider using augmented assignment instead: sum += i
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: i += 1
test/all/syntax.tact:150:9:
  149 |     while (i < 10) {
> 150 |         i = i + 1;
                ^
  151 |         sum = sum + i;
Help: Consider using augmented assignment instead: i += 1
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: sum += i
test/all/syntax.tact:151:9:
  150 |         i = i + 1;
> 151 |         sum = sum + i;
                ^
  152 |     }
Help: Consider using augmented assignment instead: sum += i
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: i -= 1
test/all/syntax.tact:155:9:
  154 |     do {
> 155 |         i = i - 1;
                ^
  156 |         sum = sum + i;
Help: Consider using augmented assignment instead: i -= 1
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: sum += i
test/all/syntax.tact:156:9:
  155 |         i = i - 1;
> 156 |         sum = sum + i;
                ^
  157 |     } until (i <= 0);
Help: Consider using augmented assignment instead: sum += i
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: i += 1
test/all/syntax.tact:160:9:
  159 |     repeat (10) {
> 160 |         i = i + 1;
                ^
  161 |         sum = sum + i;
Help: Consider using augmented assignment instead: i += 1
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: sum += i
test/all/syntax.tact:161:9:
  160 |         i = i + 1;
> 161 |         sum = sum + i;
                ^
  162 |     }
Help: Consider using augmented assignment instead: sum += i
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign

[INFO] PreferAugmentedAssign: Prefer augmented assignment: sum += value
test/all/syntax.tact:170:9:
  169 |     foreach (key, value in m) {
> 170 |         sum = sum + value;
                ^
  171 |     }
Help: Consider using augmented assignment instead: sum += value
See: https://nowarp.io/tools/misti/docs/detectors/PreferAugmentedAssign